### PR TITLE
feat: Introduce ProguardCache format

### PR DESCRIPTION
**Pull request by loewenheim**
This is a stable binary format similar to symcache. It has the exact same features as `ProguardMapper`, but doesn't need to be loaded into memory. I haven't tested performance yet.

There's also a bit of duplication, and some functions just refer to the `mapper` module.

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+        let mut frame = frame.clone();
+        let Ok(original_class) = self.read_string(class.original_name_offset) else {
+            return RemappedFrameIter::empty();
+        };
+
+        frame.class = original_class;
+
+        // The following if and else cases are very similar. The only difference
+        // is that if the frame contains parameter information, we use it in
+        // our comparisons (in addition to the method name).
```
Comment (Swatinem) [inline src/cache.rs:187]: You could potentially solve this with generics, taking a `<T: PartialEq, F: Fn(&raw::Member) -> T>`, though a bit of duplication also doesn’t hurt.

**Review by Swatinem — COMMENTED** [review]
That is some impressive work! 🎉

File: Cargo.toml [hunk 2]
```diff
@@ -14,8 +14,10 @@ edition = "2018"
 uuid = ["uuid_", "lazy_static"]
 
 [dependencies]
-uuid_ = { package = "uuid", version = "1.0.0", features = ["v5"], optional = true }
 lazy_static = { version = "1.4.0", optional = true }
+thiserror = "1.0.61"
+uuid_ = { package = "uuid", version = "1.0.0", features = ["v5"], optional = true }
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply Cargo.toml:19]: In general you get the feeling that this crate could use some love ^^

File: benches/proguard_mapping.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    let mapper = ProguardMapper::new(mapping);
+
+    let mut group = c.benchmark_group("Proguard Remapping");
+
+    group.bench_function("Cache", |b| {
+        b.iter(|| cache.remap_stacktrace(black_box(RAW)))
+    });
+    group.bench_function("Mapper", |b| {
+        b.iter(|| mapper.remap_stacktrace(black_box(RAW)))
+    });
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply benches/proguard_mapping.rs:35]: I tried that just out of interest, it doesn't go well for the mapper. Can surely add it as well.

File: src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+        let class = self.get_class(class)?;
+        let members = self.get_class_members(class)?;
+        let mut iter = members.iter().filter(|m| {
+            let Ok(obfuscated_name) = self.read_string(m.obfuscated_name_offset) else {
+                return false;
+            };
+
+            obfuscated_name == method
+        });
+        let first = iter.next()?;
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply src/cache.rs:150]: No, I don't think so. I think we'd have to do the same thing as in `remap_frame`, i.e. binary search and then linearly search forwards and backwards.

File: src/cache.rs (hunk 1 above)
Comment (loewenheim) [reply src/cache.rs:190]: The problem is I need two closures, one for the binary search and one for the linear search.

Comment (Swatinem) [reply src/cache.rs:190]: But shouldn’t `PartialEq` be a subset of `PartialOrd`? Meaning that you are only interested in `!Ordering::Equal`.

Comment (loewenheim) [reply src/cache.rs]: You're right, that works. Good thinking.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (loewenheim) [reply Cargo.toml:19]: I'll address these in a separate PR.

**Review by Swatinem — APPROVED** [review]

**Pull request merged**